From cc3835eade4265dee12faee4694ea60231c606e3 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Tue, 24 Apr 2018 13:02:36 +0800
Subject: [PATCH 3689/5242] MLK-18123-3 gpu: imx: imx8_dprc: Set
 dprc->has_aux_prg dynamically for DC0 BLIT1

commit  1303b1352ef984e34e4a073a1e25fcb9bfc4fd6d from
https://source.codeaurora.org/external/imx/linux-imx.git

With the DPR vs PRG muxing introduced in silicon, the DPR channel of
DC0 BLIT1 shares auxiliary PRG with the DPR channel of DC0 BLIT0.
When DC0 BLIT1 doesn't use the PRG, we should clear dprc->has_aux_prg,
otherwise, if DC0 BLIT0 uses the PRG as a primary one, the PRG would
be accidentally disabled by DC0 BLIT1.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/imx8_dprc.c |   18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/imx/imx8_dprc.c b/drivers/gpu/imx/imx8_dprc.c
index 536d41b..562f253 100644
--- a/drivers/gpu/imx/imx8_dprc.c
+++ b/drivers/gpu/imx/imx8_dprc.c
@@ -420,14 +420,24 @@ void dprc_configure(struct dprc *dprc, unsigned int stream_id,
 				dprc_prg_sel_configure(dprc,
 						       SC_R_DC_0_BLIT0, true);
 				prg_set_auxiliary(dprc->prgs[1]);
+				dprc->has_aux_prg = true;
 			}
 		}
 		dprc_write(dprc, uv_baddr, FRAME_2P_BASE_ADDR_CTRL0);
 	} else {
-		if (dprc->sc_resource == SC_R_DC_0_BLIT0
-			&& dprc->devtype->has_fixup) {
-			dprc_prg_sel_configure(dprc, SC_R_DC_0_BLIT0, false);
-			prg_set_primary(dprc->prgs[0]);
+		if (dprc->devtype->has_fixup) {
+			switch (dprc->sc_resource) {
+			case SC_R_DC_0_BLIT0:
+				dprc_prg_sel_configure(dprc,
+						       SC_R_DC_0_BLIT0, false);
+				prg_set_primary(dprc->prgs[0]);
+				break;
+			case SC_R_DC_0_BLIT1:
+				dprc->has_aux_prg = false;
+				break;
+			default:
+				break;
+			}
 		}
 
 		switch (modifier) {
-- 
1.7.9.5

